File: backend/src/main/java/com/doctorapp/medicaltracker/service/PrescriptionService.java
package com.doctorapp.medicaltracker.service;

import java.util.List;

import com.doctorapp.medicaltracker.dto.PrescriptionUpsertRequest;
import com.doctorapp.medicaltracker.model.Prescription;

public interface PrescriptionService {
    List<Prescription> getByCaseId(Long caseId);

    List<Prescription> getByPatientId(Long patientId);

    Prescription createDraft(Long caseId, PrescriptionUpsertRequest request);

    Prescription updateDraft(Long id, PrescriptionUpsertRequest request);

    Prescription issue(Long id);

    Prescription recordPrint(Long id);

    Prescription voidPrescription(Long id, String reason);

    void deleteDraft(Long id);
}

File: frontend/src/app/core/services/prescription.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { Prescription, PrescriptionUpsertRequest } from '../models/prescription.model';

@Injectable({ providedIn: 'root' })
export class PrescriptionService {
  private readonly apiBaseUrl = environment.apiBaseUrl;

  constructor(private readonly http: HttpClient) {}

  getByCaseId(caseId: number): Observable<Prescription[]> {
    return this.http.get<Prescription[]>(`${this.apiBaseUrl}/api/prescriptions/cases/${caseId}`);
  }

  getByPatientId(patientId: number): Observable<Prescription[]> {
    return this.http.get<Prescription[]>(`${this.apiBaseUrl}/api/prescriptions/patients/${patientId}`);
  }

  createDraft(caseId: number, request: PrescriptionUpsertRequest): Observable<Prescription> {
    return this.http.post<Prescription>(`${this.apiBaseUrl}/api/prescriptions/cases/${caseId}`, request);
  }

  updateDraft(id: number, request: PrescriptionUpsertRequest): Observable<Prescription> {
    return this.http.put<Prescription>(`${this.apiBaseUrl}/api/prescriptions/${id}`, request);
  }

  issue(id: number): Observable<Prescription> {
    return this.http.post<Prescription>(`${this.apiBaseUrl}/api/prescriptions/${id}/issue`, {});
  }

  recordPrint(id: number): Observable<Prescription> {
    return this.http.post<Prescription>(`${this.apiBaseUrl}/api/prescriptions/${id}/print`, {});
  }

  voidPrescription(id: number, reason: string): Observable<Prescription> {
    return this.http.post<Prescription>(`${this.apiBaseUrl}/api/prescriptions/${id}/void`, { reason });
  }

  deleteDraft(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiBaseUrl}/api/prescriptions/${id}`);
  }
}
